### PR TITLE
[jit] New Iterable interface, refactor for loops (List, Tensor, Range)

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21666 [jit] Add for in enumerate to script
* **#21403 [jit] New Iterable interface, refactor for loops (List, Tensor, Range)**

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -6261,6 +6261,17 @@ def fn():
         with self.assertRaisesRegex(torch.jit.Error, "must not be zero"):
             fn()
 
+    def test_script_for_in_list(self):
```
Comment (Krovatkin) [inline test/test_jit.py]: i think this is probably covered by `get_sum_list_fn` tests ( https://github.com/pytorch/pytorch/blob/master/test/test_jit.py#L3653 ) , but I might be mistaken.

**Review by Krovatkin — APPROVED** [review]
:shipit:

**Review by eellison — COMMENTED** [review]
What would it take to land this PR without duplicating all of emitLoopCommon? We shouldn't have have a `emitNewLoopCommon` and a `emitLoopCommon`? 

Could you wait to land your stack until there is only one ?

**Comment by wanchaol** [comment]
> What would it take to land this PR without duplicating all of emitLoopCommon? We shouldn't have have a `emitNewLoopCommon` and a `emitLoopCommon`?
> 
> Could you wait to land your stack until there is only one ?

@eellison Sure, `emitNewLoopCommon` is there to allow gradually transfer other for loops and easier reviews. We will only have one `emitLoopCommon` in the end, I will land the whole stack together once there's no duplicates.

File: torch/csrc/jit/script/sugared_value.h [hunk 2]
```diff
@@ -94,6 +94,32 @@ struct TORCH_API SugaredValue
     throw ErrorReport(loc) << "cannot call a " << kind();
   }
 
+  // If it is a iterable sugared value, use this interface to get the type information
+  // for all the iterables inside for loop, and prepare for loop information filling
+  virtual std::vector<TypePtr> getItersTypeInfo(
+    const SourceRange& loc,
+    Function& m) {
+   throw ErrorReport(loc) << kind() << " does not have the iterables type information";
```
Comment (zdevito) [inline torch/csrc/jit/script/sugared_value.h]: If these errors are user facing and need to describe what is happening in the same way python would if something could not be iterated.

File: torch/csrc/jit/script/sugared_value.cpp [hunk 3]
```diff
@@ -236,6 +236,54 @@ std::shared_ptr<SugaredValue> SimpleValue::call(
...
+  Graph& g = *m.graph();
+  if (auto list_type = val_type->cast<ListType>()) {
+    // fill in max_trip_count_val
+    WithInsertPoint guard(n);
+    Value* max_trip_count_val = g.insert(aten::len, {val}, {}, loc);
+    n->insertInput(0, max_trip_count_val);
+
+    // fill in the target element assignment value in the beginning of the FOR loop
+    {
+      TORCH_CHECK(iters_size == 1, "more than one iterable for in list");
```
Comment (zdevito) [inline torch/csrc/jit/script/sugared_value.cpp]: This needs to be an error report, context is needed!

**Review by zdevito — COMMENTED** [review]
Just a couple small things. What would it take to make sure we don't land extra loop emitting statements and do it in one PR? Elias is working on a lot of loop related stuff and it will slow his work down to have a lot of noise here.

**Comment by wanchaol** [comment]
> Just a couple small things. What would it take to make sure we don't land extra loop emitting statements and do it in one PR? Elias is working on a lot of loop related stuff and it will slow his work down to have a lot of noise here.

@zdevito that requires migrating the for_range and while together, it requires extra work on hooking up Apply expressions, but it shouldn't be a problem. I will try to do them in this PR to avoid duplicating.

**Comment by zdevito** [comment]
Let me know when the version working with the new Loop is ready to go. Maybe coordinate with Elias, since https://github.com/pytorch/pytorch/pull/21611 will also affect this logic.

**Comment by wanchaol** [comment]
close in favor of https://github.com/pytorch/pytorch/pull/21801

**Pull request closed without merging**